File: src/main/java/com/hileco/cortex/vm/ProgramContext.java
package com.hileco.cortex.vm;

import com.hileco.cortex.vm.layer.LayeredBytes;
import com.hileco.cortex.vm.layer.LayeredStack;
import lombok.Data;

import java.math.BigInteger;

@Data
public class ProgramContext {
    public static final int INSTRUCTION_LIMIT = 1_000_000;

    private int instructionsExecuted;
    private int instructionLimit;
    private int instructionPosition;
    private LayeredStack<byte[]> stack;
    private LayeredBytes memory;
    private Program program;
    private BigInteger returnDataOffset;
    private BigInteger returnDataSize;
    private LayeredBytes callData;

    public ProgramContext(Program program) {
        this.instructionsExecuted = 0;
        this.instructionPosition = 0;
        this.instructionLimit = INSTRUCTION_LIMIT;
        this.stack = new LayeredStack<>();
        this.memory = new LayeredBytes();
        this.callData = new LayeredBytes();
        this.returnDataOffset = BigInteger.ZERO;
        this.returnDataSize = BigInteger.ZERO;
        this.program = program;
    }
}

File: src/main/java/com/hileco/cortex/vm/ProgramContext.kt
package com.hileco.cortex.vm

import com.hileco.cortex.vm.layer.LayeredBytes
import com.hileco.cortex.vm.layer.LayeredStack
import java.math.BigInteger

data class ProgramContext(val program: Program,
                          var instructionsExecuted: Int = 0,
                          var instructionPosition: Int = 0,
                          val instructionLimit: Int = INSTRUCTION_LIMIT,
                          val stack: LayeredStack<ByteArray> = LayeredStack(),
                          val memory: LayeredBytes = LayeredBytes(),
                          var returnDataOffset: BigInteger = BigInteger.ZERO,
                          var returnDataSize: BigInteger = BigInteger.ZERO,
                          val callData: LayeredBytes = LayeredBytes()) {
    companion object {
        const val INSTRUCTION_LIMIT = 1000000
    }
}